### PR TITLE
refactor: crazy breaking refactor

### PR DESCRIPTION
**Pull request by P403n1x87**
This refactor introduces the concept of Tasks. Key change is the parenting of venv instances which allows determining the `PYTHONPATH` to use when running commands inside nested venvs.

## Potential future improvements

- Add subtasks to tasks
- Add dependencies between tasks

### CODE REVIEW

File: riot/venv.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        return " ".join(
+            [
+                f"'{get_pep_dep(lib, version)}'"
+                for lib, version in self.pkgs
+                if version is not None
+            ]
+        )
+
+    @property
+    def pythonpath(self):
```
Comment (P403n1x87) [inline riot/venv.py:73]: Don't mind the debug noise. By walking up the parenting relation we can determine the PYTHONPATH to use in order to import the correct dependencies 🤞

File: riot/venv.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    ) -> t.Generator[Venv, None, None]:
+        pys = pys or self.pys or [None]
+        for py in pys:
+            for ps in expand_specs(self.pkgs):
+                venv = Venv(
+                    py=py,
+                    pkgs=ps,
+                    parent=parent_venv,
+                )
+                if not self.venvs:
```
Comment (P403n1x87) [inline riot/venv.py:198]: I think the `if` and the `for` loops should be this way (currently it's inside-out). This way one can set the parenting relation between venv instances.

File: riot/__init__.py [hunk 3]
```diff
@@ -1,5 +1,7 @@
-__all__ = ["latest", "Session", "Venv"]
+__all__ = ["latest", "Session", "Task", "Venv"]
 
-from .riot import Session, Venv
+from riot.riot import Session
+from riot.task import Task
+from riot.venv import VenvSpec as Venv
```
Comment (P403n1x87) [inline riot/__init__.py:5]: If the argument `venvs` of `Task` could be made of type `List[Dict[..., ...]]` then `Venv` wouldn't be required to be part of the public API.

**Review by Kyle-Verhoog — COMMENTED** [review]
Interestingly enough this brings us roughly back to the early days of riot where it was `Suite` and `Venv` where `Suite` defined the name and command IIRC. Now it's called `Task` which I do like a bit better.

I really like the idea of having a defined top level structure which makes the riotfile more intuitive than all the nested Venv things. Also helps clean up the implementation because we don't need to worry about so many things down the tree.

We could do some of the file break out changes ahead of time which I think would help clean this PR up some.

I'm personally not a fan of having a `utils` file as it just becomes a dumping ground and doesn't really save much clutter. Not a super strong opinion though.


Overall good stuff and I think this is a good improvement!

File: riot/venv.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+class VenvError(Exception):
+    pass
+
+
+@dataclasses.dataclass
+class Venv:
+    pkgs: t.Tuple[t.Tuple[str, str]]
+    py: t.Optional[Interpreter]
+    parent: t.Optional["Venv"] = None
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply riot/venv.py:30]: But tasks can share a venv :)

~~~ python
checks_venv = Venv(
    pkgs = { "black": latest, "isort": latest }
)

tasks = [
    Task(name="fmt", command="black . && isort .", venvs=[checks_venv])
    Task(name="black", command="black .", venvs=[checks_venv])
    Task(name="isort", command="isort .", venvs=[checks_venv])
]
~~~

[the remark this answers is not in the document]
Comment (Kyle-Verhoog) [reply riot/venv.py:30]: oh sure, although it does become more procedural and less declarative/json/yaml-able

**Comment by P403n1x87** [comment]
The main branch has diverged considerably. We will maybe take the general idea and implement it with a new PR in the future, but we might split the `riot.py` source first.

**Pull request closed without merging**